### ISSUE TITLE
Bug: [no-unnecessary-type-conversion] Unsafe suggestion: removing bitwise `~~` changes semantics for non-integer numbers

### ISSUE THREAD

**Issue by ziloen** [typescript-eslint/typescript-eslint#11491]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.9.2&fileType=.tsx&code=MYewdgzgLgBAhjAvDAjABgHQoLACg%2BiSwBGSMAfuXHgPQ0wMwB6qaQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWmQ5MAbomjJKM9FAXQO0SODABfELqA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
const a = 1.1
const b = ~~a
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unnecessary-type-conversion": "error"
  }
};
```

### tsconfig

```jsonc

```

### Expected Result

No suggestion for `~~` case

### Actual Result

The rule reports:
> Using ~~ on a number does not change the type or value of the number.

**Comment by kirkwaiblinger** [comment]
So, is the solution:

a) only report on numeric literal types, like the following, and keep the same report messages?
```ts
declare const threeOrFour: 3 | 4;
const alreadyNumber = ~~threeOrFour
```

b) remove the `~~` check altogether?

c) keep the report, but remove the suggestion for removing `~~`. Possibly mention `Math.trunc()`/`Math.floor()` instead?

**Comment by bradzacher** [comment]
Probably (a)? We know it's a useless operator in that case.
I don't think it's our place to do (c) -- that's a purely stylistic choice and doesn't fit into the rule

**Comment by mouse0429** [comment]
Hi! I'm working on implementing the fix for this issue. Currently, my implementation only handles **literal number types** as mentioned in the discussion above (option a), but I found there's an existing test case that expects broader behavior.

**Current implementation (literal number types only):**
- `~~5` → Reports as unnecessary ✅  
- `declare const x: 3 | 4; ~~x` → Reports as unnecessary ✅
- `~~1.1` → Not reported (decimal literal) ✅

**Existing test case that fails with current approach:**
- `2 * ~~(2 + 2);` → Expected to report `~~(2 + 2)` as unnecessary, but `(2 + 2)` is an expression, not a literal type

**Question:** Should we:
1. **Expand the rule** to handle expressions/function calls that evaluate to integers:
   - `~~(2 + 2)` (arithmetic expressions)
   - `~~Number('3')` (function calls)
   - Other statically analyzable integer values

2. **Remove the conflicting test case** and stick to literal number types only as originally discussed

**Comment by kirkwaiblinger** [comment]
@mouse0429 Nope, `2 + 2` is just a black box with TS type `number` for all we know! Generally we do not try to "do better" than the TS types unless there is a very strong motivation for it.

**Comment by injae-kim** [comment]
Fix PR: https://github.com/typescript-eslint/typescript-eslint/pull/11517 PTAL 😄
